unit_tests/test_redis_client_base.py

```diff
@@ -0,0 +1,237 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import json
+import uuid
+from unittest.mock import MagicMock, patch
+
+import pytest
+from redis import StrictRedis
+
+import d3a_api_client
+from d3a_api_client.redis_client_base import RedisClientBase, RedisAPIException
+
+AREA_ID = str(uuid.uuid4())
+TRANSACTION_ID = str(uuid.uuid4())
+DEVICE_ID = str(uuid.uuid4())
+
+
+# pylint: disable=protected-access
+class TestRedisClientBase:
+    """Tests for the redis client base behaviour without connecting to d3a."""
+
+    @staticmethod
+    @pytest.fixture()
+    def redis_client_auto_register():
+        """Create the fixture for redis client base and mocks strict redis."""
+        d3a_api_client.redis_client_base.StrictRedis = MagicMock(spec=StrictRedis)
```

Comment:
Please be aware that this is not the proper way to mock module-level objects (`StrictRedis` in this case). As you've done elsewhere in your code, the correct way to do this is to use `patch()`.

In our case there's one more issue: we don't want to just directly `RedisClientBase().redis_db`, because we need to change its value during the instance initialization (not afterwards). That's because `RedisClientBase().pubsub` depends on it during the init stage.

So my suggestion is to patch the `StrictRedis.from_url` method (in the fixture) and return a mock from it. If you do so, you'll be able to delete all these lines from the other methods:
```python
redis_client_auto_register.redis_db.publish = MagicMock()
```

Reply:
I think you are correct but I tried the fixture with a patch. Every unit test function needs the Redis connection in this case. Due to this, I have to use the return value of the fixture in all the functions. If I can mock the Strict Redis in the redis_auto_register then I don't have to mock any Redis-related functions again. For this case we can just MagicMock StrictRedis once in the fixture and avoid other magic mocks for redis.

Reply:
That's exactly what I was suggesting ;)

Reply:
I tried to mock and return the value of from_url but the behavior hasn't changed anything even just mocking only StrictRedis. So I just mocked the StrictRedis in the fixture and removed other lines of mocking redis_db for other functions. Now iit is working fine.